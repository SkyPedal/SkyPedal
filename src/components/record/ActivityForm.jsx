import { useEffect, useState } from "react";
import DistanceDuration from "./DistanceDuration";
import PropTypes from "prop-types";
import StartEnd from "./StartEnd";
import TabSelector from "./TabSelector";
import DragDrop from "./DragDrop";
import useFormData from "./FormData";

const ActivityForm = ({ handleSave, api, setError }) => {
  const [tab, setTab] = useState("startEnd");
  const [reset, setReset] = useState(false);
  const formData = useFormData();

  useEffect(() => {
    if (reset) {
      formData.reset();
      setReset(false);
    }
  }, [reset, formData]);

  const changeTab = (tab) => {
    setTab(tab);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    console.log("IM SUBMITTING");
    handleSave({
      title: formData.data.title,
      date: formData.data.date,
      time: formData.data.time,
      distance: formData.data.distance,
      duration: formData.data.duration,
      geoJson: formData.data.geoJson,
    });
  };

  const canTabChange =
    (tab === "startEnd" &&
      formData.data.start === "none" &&
      formData.data.end === "none") ||
    (tab === "distanceDuration" &&
      formData.data.distance === 0 &&
      formData.data.duration === 0) ||
    (tab === "dragDrop" && formData.data.geoJson === null);

  return (
    <form onSubmit={handleSubmit} className="h-[90%]">
      <div className="flex h-full flex-col pt-12">
        <label htmlFor="activityName" className="text-left text-lg">
          Title
        </label>
        <input
          type="text"
          id="activityName"
          name="activityName"
          className="rounded-lg border-2 border-solid border-gray-300 p-2"
          value={formData.data.title}
          onChange={(e) => formData.setField("title", e.target.value)}
        />
        <div className="flex flex-row justify-between">
          <div className="flex flex-col">
            <label htmlFor="activityDate" className="pt-5 text-left text-lg">
              Date
            </label>
            <input
              type="date"
              id="activityDate"
              name="activityDate"
              className="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              value={formData.data.date}
              onChange={(e) => formData.setField("date", e.target.value)}
            />
          </div>
          <div className="flex flex-col">
            <label htmlFor="activityTime" className="pt-5 text-left text-lg">
              Time
            </label>
            <input
              type="time"
              id="activityTime"
              name="activityTime"
              className="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              value={formData.data.time}
              onChange={(e) => formData.setField("time", e.target.value)}
            />
          </div>
        </div>
        {/* Tab Content */}
        <div className="mt-5 flex h-[65%] flex-col justify-between border-y-2 py-2">
          <div className="flex-grow pt-2">
            <TabSelector
              currentTab={tab}
              changeTab={changeTab}
              lock={!canTabChange}
            />
            {tab === "startEnd" && (
              <StartEnd formData={formData} api={api} setError={setError} />
            )}
            {tab === "distanceDuration" && (
              <DistanceDuration formData={formData} reset={reset} />
            )}
            {tab === "dragDrop" && (
              <DragDrop reset={reset} formData={formData} />
            )}
          </div>

          <div className="left-4 flex flex-row justify-start pb-2">
            <p
              data-cy="distance-chip"
              className="mt-5 rounded-lg border-2 border-solid border-gray-300 bg-gray-100 px-4 py-2"
            >
              {(formData.data.distance / 1000).toFixed(2)} km
            </p>
            <p
              data-cy="duration-chip"
              className="mx-2 mt-5 rounded-lg border-2 border-solid border-gray-300 bg-gray-100 px-4 py-2"
            >
              {Math.round(formData.data.duration / 60)} mins
            </p>
            {formData.data.geoJson ? (
              <p
                data-cy="gps-chip"
                className="mx-2 mt-5 rounded-lg border-2 border-solid border-gray-300 bg-green-100 px-4 py-2"
              >
                GPS
              </p>
            ) : (
              <p
                data-cy="gps-chip"
                className="mx-2 mt-5 rounded-lg border-2 border-solid border-gray-300 bg-gray-100 px-4 py-2"
              >
                No GPS
              </p>
            )}
          </div>
        </div>
      </div>

      {/* Buttons  */}
      <div className="absolute bottom-12 right-4 flex w-full flex-row justify-end">
        <button
          name="clear"
          type="button"
          className="mx-8 mt-5 max-h-10 w-24 rounded bg-red-500 px-4 py-2 font-bold text-white hover:bg-red-700"
          onClick={() => setReset(true)}
        >
          Clear
        </button>
        <button
          name="save"
          type="submit"
          className="mt-5 max-h-10 w-24 rounded bg-blue-500 px-4 py-2 font-bold text-white hover:bg-blue-700 disabled:bg-red-300"
          disabled={
            formData.data.distance === 0 || formData.data.duration === 0
              ? true
              : false
          }
        >
          Add
        </button>
      </div>
    </form>
  );
};

PropTypes.ActivityForm = {
  handleSave: PropTypes.func.isRequired,
  locations: PropTypes.array.isRequired,
  formData: {
    data: PropTypes.object.isRequired,
    setField: PropTypes.func.isRequired,
    set: PropTypes.func.isRequired,
    reset: PropTypes.func.isRequired,
  },
  setFormData: PropTypes.func.isRequired,
};

export default ActivityForm;
