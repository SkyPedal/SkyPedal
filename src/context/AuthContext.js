import { createContext, useContext } from "react";

const AuthContext = createContext();

export default AuthContext;

export const useAuth = () => {
  return useContext(AuthContext);
};
