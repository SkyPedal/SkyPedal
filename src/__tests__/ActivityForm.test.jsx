import { act, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { describe, expect, it, vi } from "vitest";
import ActivityForm from "../components/record/ActivityForm";
import { BrowserRouter as Router } from "react-router-dom";
import AuthContext from "../context/AuthContext";
import sampleData from "../../database.sample";

const sampleFormData = {
  title: "Commute to Work",
  date: "2024-01-01",
  time: "08:00",
  start: "none",
  end: "none",
  distance: 0,
  duration: 0,
};

describe("ActivityForm", () => {
  it("snapshot renders", () => {
    const { asFragment } = render(
      <Router>
        <AuthContext.Provider value={{ user_id: 1, user_name: "TestUser123" }}>
          <ActivityForm
            handleSave={() => null}
            locations={sampleData.locations}
            formData={sampleFormData}
            setFormData={() => null}
          />
        </AuthContext.Provider>
      </Router>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders form with default values", () => {
    const { getByLabelText, getByText } = render(
      <Router>
        <AuthContext.Provider value={{ user_id: 1, user_name: "TestUser123" }}>
          <ActivityForm
            handleSave={() => null}
            locations={sampleData.locations}
            formData={sampleFormData}
            setFormData={() => null}
          />
        </AuthContext.Provider>
      </Router>,
    );
    expect(getByLabelText("Title").value).toBe("Commute to Work");
    expect(getByLabelText("Date").value).toBe("2024-01-01");
    expect(getByLabelText("Time").value).toBe("08:00");
    expect(getByLabelText("Start Location").value).toBe("none");
    expect(getByLabelText("End Location").value).toBe("none");
    expect(getByLabelText("Distance (m)").value).toBe("");
    expect(getByLabelText("Duration (s)").value).toBe("");
    expect(getByText("Cancel")).toBeInTheDocument();
    expect(getByText("Add")).toBeInTheDocument();
  });

  it("should disable the distance/duration when locations are set", () => {
    const setFormData = vi.fn();
    const { getByLabelText } = render(
      <Router>
        <AuthContext.Provider value={{ user_id: 1, user_name: "TestUser123" }}>
          <ActivityForm
            handleSave={() => null}
            locations={sampleData.locations}
            formData={sampleFormData}
            setFormData={setFormData}
          />
        </AuthContext.Provider>
      </Router>,
    );
    expect(getByLabelText("Distance (m)")).not.toBeDisabled();
    expect(getByLabelText("Duration (s)")).not.toBeDisabled();
    fireEvent.change(getByLabelText("Start Location"), {
      target: { value: "1" },
    });
    expect(setFormData).toHaveBeenCalledWith({
      ...sampleFormData,
      start: "1",
    });
  });
  it("should disable the distance/duration when start set", () => {
    const setFormData = vi.fn();
    const { getByLabelText } = render(
      <Router>
        <AuthContext.Provider value={{ user_id: 1, user_name: "TestUser123" }}>
          <ActivityForm
            handleSave={() => null}
            locations={sampleData.locations}
            formData={{ ...sampleFormData, start: "1" }}
            setFormData={setFormData}
          />
        </AuthContext.Provider>
      </Router>,
    );
    expect(getByLabelText("Distance (m)")).toBeDisabled();
    expect(getByLabelText("Duration (s)")).toBeDisabled();
  });
  it("should disable the distance/duration when end set", () => {
    const setFormData = vi.fn();
    const { getByLabelText } = render(
      <Router>
        <AuthContext.Provider value={{ user_id: 1, user_name: "TestUser123" }}>
          <ActivityForm
            handleSave={() => null}
            locations={sampleData.locations}
            formData={{ ...sampleFormData, end: "1" }}
            setFormData={setFormData}
          />
        </AuthContext.Provider>
      </Router>,
    );
    expect(getByLabelText("Distance (m)")).toBeDisabled();
    expect(getByLabelText("Duration (s)")).toBeDisabled();
  });
});
