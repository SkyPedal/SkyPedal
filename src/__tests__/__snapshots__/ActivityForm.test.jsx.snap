// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityForm > snapshot renders 1`] = `
<DocumentFragment>
  <form>
    <div
      class="flex flex-col pt-12"
    >
      <label
        class="text-left text-lg"
        for="activityName"
      >
        Title
      </label>
      <input
        class="rounded-lg border-2 border-solid border-gray-300 p-2"
        id="activityName"
        name="activityName"
        type="text"
        value="Commute to Work"
      />
      <div
        class="flex flex-row justify-between"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="pt-5 text-left text-lg"
            for="activityDate"
          >
            Date
          </label>
          <input
            class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
            id="activityDate"
            name="activityDate"
            type="date"
            value="2024-01-01"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="pt-5 text-left text-lg"
            for="activityTime"
          >
            Time
          </label>
          <input
            class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
            id="activityTime"
            name="activityTime"
            type="time"
            value="08:00"
          />
        </div>
      </div>
      <div
        class="mt-12"
      >
        <div
          class="flex flex-row justify-between"
        >
          <div
            class="flex flex-col"
          >
            <label
              class="pt-5 text-left text-lg"
              for="activityStart"
            >
              Start Location
            </label>
            <select
              class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              id="activityStart"
              name="activityStart"
            >
              <option
                value="none"
              >
                -- Pick --
              </option>
              <option
                value="1"
              >
                Home
              </option>
              <option
                value="2"
              >
                Work - Watermark
              </option>
              <option
                value="Add New"
              >
                Add New
              </option>
            </select>
          </div>
          <div
            class="flex flex-col"
          >
            <label
              class="pt-5 text-left text-lg"
              for="activityEnd"
            >
              End Location
            </label>
            <select
              class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              id="activityEnd"
              name="activityEnd"
            >
              <option
                value="none"
              >
                -- Pick --
              </option>
              <option
                value="1"
              >
                Home
              </option>
              <option
                value="2"
              >
                Work - Watermark
              </option>
              <option
                value="Add New"
              >
                Add New
              </option>
            </select>
          </div>
        </div>
      </div>
      <p
        class="mb-2 mt-8 w-full border-b-2 text-center"
        style="line-height: 0.1em;"
      >
        <span
          class="bg-white pl-2 pr-2"
        >
          OR
        </span>
      </p>
      <div
        class="flex flex-row justify-between"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="pt-5 text-left text-lg"
            for="activityDistance"
          >
            Distance (m)
          </label>
          <input
            class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
            id="activityDistance"
            min="1"
            name="activityDistance"
            required=""
            type="number"
            value=""
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="pt-5 text-left text-lg"
            for="activityDuration"
          >
            Duration (s)
          </label>
          <input
            class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
            id="activityDuration"
            min="1"
            name="activityDuration"
            required=""
            type="number"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="absolute bottom-12 right-4 flex w-full flex-row justify-end"
    >
      <button
        class="mx-8 mt-5 max-h-10 w-24 rounded bg-red-500 px-4 py-2 font-bold text-white hover:bg-red-700"
        type="button"
      >
        Cancel
      </button>
      <button
        class="mt-5 max-h-10 w-24 rounded bg-blue-500 px-4 py-2 font-bold text-white hover:bg-blue-700 disabled:bg-red-300"
        disabled=""
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</DocumentFragment>
`;
